probe-rs/src/flashing/flasher.rs

```diff
@@ -244,44 +222,289 @@ impl<'a> Flasher<'a> {
+
+            // If we encounter an error, catch it, gracefully report the failure and return the error.
+            if result.is_err() {
+                progress.failed_filling();
+                return result;
+            }
+        }
+
+        // We successfully finished filling.
+        progress.finished_filling();
+
+        // Erase all necessary sectors.
+        progress.started_erasing();
 
+        if do_chip_erase {
+            self.chip_erase(&flash_layout, progress)?;
+        } else {
+            self.sector_erase(&flash_layout, progress)?;
+        }
+
+        // Flash all necessary pages.
+
+        if self.double_buffering_supported() && enable_double_buffering {
+            self.program_double_buffer(&flash_layout, progress)?;
+        } else {
+            self.program_simple(&flash_layout, progress)?;
+        };
+
+        Ok(())
+    }
+
+    /// Fills all the bytes of `current_page`.
+    ///
+    /// If `restore_unwritten_bytes` is `true`, all bytes of the page,
+    /// that are not to be written during flashing will be read from the flash first
+    /// and written again once the page is programmed.
+    pub(super) fn fill_page(
+        &mut self,
+        current_page: &mut FlashPage,
+        restore_unwritten_bytes: bool,
+    ) -> Result<(), FlashError> {
+        // The remaining bytes to be filled in at the end of the page.
+        let remaining_bytes = current_page.size as usize - current_page.data.len();
+        if current_page.data.len() != current_page.size as usize {
+            let address_remaining_start = current_page.address + current_page.data.len() as u32;
+
+            // Fill up the page with current page bytes until it's full.
+            let old_data = if restore_unwritten_bytes {
+                // Read all the remaining old bytes from flash to restore them later.
+                let mut data = vec![0; remaining_bytes];
+                self.run_verify(|active| {
+                    active.read_block8(address_remaining_start, data.as_mut_slice())
+                })?;
+                data
+            } else {
+                // Set all the remaining bytes to their default erased value.
+                vec![self.flash_algorithm().flash_properties.erased_byte_value; remaining_bytes]
+            };
+            current_page.data.extend(old_data);
+        }
         Ok(())
     }
+
+    /// Erase the entire flash of the chip.
+    ///
+    /// This takes the list of available sectors only for progress reporting reasons.
+    /// It does not indeed erase single sectors but erases the entire flash.
+    fn chip_erase(
+        &mut self,
+        flash_layout: &FlashLayout,
+        progress: &FlashProgress,
+    ) -> Result<(), FlashError> {
+        progress.started_erasing();
+
+        let mut t = std::time::Instant::now();
+        let result = self.run_erase(|active| active.erase_all());
+        for sector in flash_layout.sectors() {
```

Comment:
I think this will just show no progress while the erase is happening, and then it will instantly fill the whole bar. Maybe it would be better to have a special progress method for this?

Reply:
This is indeed the case. but we have no feedback how far the erase process has progressed. So I don't know how to do this better. Maybe if we estimated from the known chip erase T we could display an estimate, but I'd actually prefer the current implementation over that, appart from that we don't have proper erase T information yet.

Reply:
My idea would be to show a progress bar which is just spinning, just make it clear that something actually happens. 
I would prefer this to showing no progress for several seconds. If we use the same event, the user has no change to show some message indicating that a full chip erase is in progress.